### PR TITLE
feat: Implement logic for publish from node

### PR DESCRIPTION
**Pull request by alaibe**
Publish a message with retry to lightpush or relay depending on the number of peers

### CODE REVIEW

File: waku/v2/node/wakunode2.go [hunk 1]
```diff
@@ -370,9 +373,28 @@ func (w *WakuNode) Broadcaster() v2.Broadcaster {
 	return w.bcaster
 }
 
-func (w *WakuNode) mountRelay(opts ...pubsub.Option) error {
+func (w *WakuNode) Publish(msg *pb.WakuMessage) error {
+	hash, _ := msg.Hash()
+
+	err := try.Do(func(attempt int) (bool, error) {
+		var err error
+		if w.lightPush.IsStarted() || !w.relay.EnoughPeersToPublish() {
```
Comment (alaibe) [inline waku/v2/node/wakunode2.go:381]: This logic seems wrong to me. As we have no guarantee relay has been started.
wdyt?

File: waku/v2/node/wakunode2.go [hunk 2]
```diff
@@ -370,9 +373,28 @@ func (w *WakuNode) Broadcaster() v2.Broadcaster {
 	return w.bcaster
 }
 
-func (w *WakuNode) mountRelay(opts ...pubsub.Option) error {
+func (w *WakuNode) Publish(msg *pb.WakuMessage) error {
```
Comment (richard-ramos) [inline waku/v2/node/wakunode2.go:376]:
```suggestion
func (w *WakuNode) Publish(ctx context.Context, msg *pb.WakuMessage) error {
```

File: waku/v2/node/wakunode2.go [hunk 3]
```diff
@@ -370,9 +373,28 @@ func (w *WakuNode) Broadcaster() v2.Broadcaster {
...
+	hash, _ := msg.Hash()
+
+	err := try.Do(func(attempt int) (bool, error) {
+		var err error
+		if w.lightPush.IsStarted() || !w.relay.EnoughPeersToPublish() {
+			log.Debug("publishing message via lightpush", hexutil.Encode(hash))
+			_, err = w.Lightpush().Publish(w.ctx, msg)
```
Comment (richard-ramos) [inline waku/v2/node/wakunode2.go:383]:
```suggestion
			_, err = w.Lightpush().Publish(ctx, msg)
```

File: waku/v2/node/wakunode2.go [hunk 4]
```diff
@@ -370,9 +373,28 @@ func (w *WakuNode) Broadcaster() v2.Broadcaster {
...
+		} else {
+			log.Debug("publishing message via relay", hexutil.Encode(hash))
+			_, err = w.Relay().Publish(w.ctx, msg)
```
Comment (richard-ramos) [inline waku/v2/node/wakunode2.go:386]:
```suggestion
			_, err = w.Relay().Publish(ctx, msg)
```

File: waku/v2/node/wakunode2.go [hunk 5]
```diff
@@ -370,9 +373,28 @@ func (w *WakuNode) Broadcaster() v2.Broadcaster {
...
+		}
+
+		return attempt < 5, err
```
Comment (richard-ramos) [inline waku/v2/node/wakunode2.go:389]: Lets use a constant instead of the 5!

**Review by richard-ramos — COMMENTED** [review]
Very cool!

**Comment by alaibe** [comment]
@richard-ramos PR updated, with a fix on the logic for publishing message and a new WithWakuRelay accepting a default parameter

**Review by richard-ramos — APPROVED** [review]

**Pull request merged**